include/pybind11/detail/common.h

```diff
@@ -171,7 +171,18 @@
 #define PYBIND11_PLUGIN_IMPL(name) \
     extern "C" PYBIND11_EXPORT PyObject *PyInit_##name()
 
+#define PYBIND11_CHECK_REIMPORT(m)                                             \
+    {                                                                          \
+        static int module_key;                                                 \
+        auto& modules = pybind11::get_or_create_shared_data<                   \
+            std::unordered_map<int*, pybind11::handle>>("__pybind11_modules"); \
+        pybind11::handle& original = modules[&module_key];                     \
+        if (original) return original.inc_ref().ptr();                         \
+        original = m;                                                          \
```

Comment:
Uncertain myself, but asking anyway: could it make sense to move this code to an inline function in the `pybind11::detail` namespace?
It would also be nice to add a terse comment (here or in the function) with a hint why this is needed, so that people don't have to dig up this PR to find out.

Reply:
Yeah, I can delegate most of this code to a function, and could add a comment, pending other discussions - thanks!